keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+  multiple kinds, e.g. "io.k8s.api.core.v1.PodTemplateSpec" or
+  "io.k8s.api.core.v1.Container" would help policy authors apply policies more
+  broadly and uniformly (e.g. to match initContainers, containers,
+  ephemeralContainers). Note that a list of field scopes described below can
+  handle some cases like this quite well (although each place a type exists must
+  be explicitly listed).
+
+For example, to validate all containers:
+
+```yaml
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      scopes: ["spec.containers[*]"]
+```
+
+This could be extended to validate all kinds of containers in a pod if scopes
+supports a homogeniously typed list:
+
+```yaml
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      scopes: ["spec.containers[*]", "initContainers[*]", "spec.ephemeralContainers[*]"]
+```
+
+`ephemeralContainers` might not exist due to version skew (it's a relatively new
+field), in which case that scope would be dropped (but a warning would be added
+to the status).
+
+#### Secondary Authz
+
+<<[UNRESOLVED jpbetz, deads2k, tallclair, ?? ]>>
+We have general agreement to include this as a feature, but need to provide a
+more concrete design.
+<<[/UNRESOLVED]>>
+
+kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
+as a way of doing things like:
+
+- Validate that only a user with a specific permission can set a particular
+  field.
+- Validate that only a controller responsible for a finalizer can remove it from
+  the finalizers field.
+
+This could be supported by matching criteria, or via CEL expression access, or both.
+ 
+Use cases:
+
+- PodSecurityPolicy (kube)
+- CertificateApproval (kube)
+- CertificateSigning (kube)
+- OwnerReferencesPermissionEnforcement (kube)
+- network.openshift.io/ExternalIPRanger
+- route.openshift.io/IngressAdmission
+- scheduling.openshift.io/PodNodeConstraints
+- network.openshift.io/RestrictedEndpointsAdmission
+- security.openshift.io/SecurityContextConstraint
+- security.openshift.io/SCCExecRestrictions
+
+From deads2k:
+
+> Note that user.Extra in AdmissionReview has pod claims, which are valuable.
+
+> sig-auth has previous talked about trying to find a way to restrict access
+> from a daemonset pod to a customresource/foo that has Foo.spec.NodeName set to
+> the Node.metadata.name of the pod bound to the particular SA token. This is
+> tantalizingly close because user.Extra contains
+> authentication.kubernetes.io/pod-uid to locate a pod, determine a
+> Pod.spec.NodeName.
+
+> A built-in that does that may be well received and unlock many use-cases.
+> Exploring the idea may be useful. If most also require controlled read
+> permission, then its probably better to create something specifically for the
+> purpose.
+
+Looking up the pod (or any other additional resources) is not something we are
+currently planning to support in this KEP, but the use case is interesting and
+we should investigate with sig-auth.
```

Comment:
Agreed.  This was a CVE for numerous distributions.

Reply:
+1